### PR TITLE
Code polishing of Prune(TableScanColumns, ValuesColumns)

### PR DESCRIPTION
**Pull request by kokosing**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneTableScanColumns.java [hunk 1]
```diff
@@ -72,11 +69,11 @@ public Pattern getPattern()
                                 child.getId(),
                                 child.getTable(),
                                 newOutputs,
-                                newOutputs.stream()
-                                        .collect(Collectors.toMap(Function.identity(), e -> child.getAssignments().get(e))),
+                                filterKeys(child.getAssignments(), newOutputs::contains),
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneTableScanColumns.java]: The only purpose of newOutputs is to preserve the previous order of outputs to not change the node unnecessarily.  Using dependencies.get()::contains in filterKeys would in my opinion be more direct and would scale properly with the number of columns.

File: presto-main/src/main/java/com/facebook/presto/util/MoreLists.java [hunk 2]
```diff
@@ -29,5 +30,12 @@
                 .collect(toImmutableList());
     }
 
+    public static <T> List<T> filter(List<T> elements, Predicate<T> predicate)
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/util/MoreLists.java]: I like this.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneJoinColumns.java [hunk 3]
```diff
@@ -73,9 +73,7 @@ public Pattern getPattern()
                                 joinNode.getLeft(),
                                 joinNode.getRight(),
                                 joinNode.getCriteria(),
-                                joinNode.getOutputSymbols().stream()
-                                        .filter(dependencies.get()::contains)
-                                        .collect(toImmutableList()),
+                                filter(joinNode.getOutputSymbols(), dependencies.get()::contains),
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneJoinColumns.java]: Great!

File: presto-main/src/main/java/com/facebook/presto/util/MoreLists.java [hunk 4]
```diff
@@ -29,5 +30,12 @@
...
+    {
+        return elements.stream()
+                .filter(predicate)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/util/MoreLists.java]: Consider checking nullity explicitly.

File: presto-main/src/main/java/com/facebook/presto/util/MoreLists.java (hunk 2 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/util/MoreLists.java]: Does it matter that `elements` is a list? Can it be any `Collection`? Or `Iterable` (perhaps with the help of `Strems.stream(Iterable)`)?

I have also a problem whenever a see `filter(list, ...)` in the code. Is it filtered copy (your new method) or filtered view (`Iterables.filter`, the only candidate I knew previously).
Can the method be named `filteredCopy`? I know this is more verbose, but it's _not_ going to be used in `filteredCopy(transformedCopy(filteredCopy(...)))` cases, streams are for that.

**Comment by alandpost** [comment]
@kokosing Is this PR not going to get merged, due to @findepi's objection to its non-stream-ness?

**Comment by kokosing** [comment]
Everything is just fine. I just did not have time to apply comments and
merge. I will try to do it today.

06.07.2017 8:03 PM "Alan Post" <notifications@github.com> napisał(a):

**Comment by findepi** [comment]
@alandpost, I wouldn't expect my objections can block a PR ;)

**Pull request merged**